doc/factors.md

```diff
@@ -0,0 +1,208 @@
+
+ ```
+ # factors
+
+_lemma
+
+_c
+c0 : _c
+c1 : _c
+c2 : _c
+_has_c
+
+_s
+s0 : _s
+s1 : _s
+_has_s
+
+ # lemmas
+
+</s> : _lemma
+<unk> : _lemma
+, : _lemma _has_c _has_s
+. : _lemma _has_c _has_s
+the : _lemma _has_c _has_s
+for: _lemma _has_c _has_s
+ ```
+
+## Training options
+
+There are two choices for how factor embeddings are combined with *lemma* embeddings: summation and concatenation.
+
+```
+--factors-combine TEXT=sum      How to combine the factors and lemma embeddings.
+                                Options available: sum, concat
+```
+
+The dimension of the factor embeddings must be specified if using combine option `concat`. If using `sum`, the factor embedding dimension matches that of the lemmas.
+
+```
+--factors-dim-emb INT           Embedding dimension of the factors. Only used if concat is selected as factors combining form
+```
+
+Note: At the moment `concat` is only implemented for usage in the source side.
+
+### Prediction
+
+If using factors on the target side, there are multiple options for how factor predictions are generated related to the form of conditioning / dependencies of factors and lemmas. If no option is set with `--lemma-dependency`, the default behavior will be predicting the factors with no lemma dependency.
+
+```
+--lemma-dependency TEXT         Lemma dependency method to use when predicting target factors.
+                                Options: soft-transformer-layer, hard-transformer-layer, lemma-dependent-bias, re-embedding
+
+--lemma-dim-emb INT=0           Re-embedding dimension of lemma in factors
+```
+
+* `soft-transformer-layer`: Uses an additional transformer layer to predict the factors using the previously predicted lemma
+* `hard-transformer-layer`: Like `soft-transformer-layer` but with hard-max
+* `lemma-dependent-bias`: Adds a learned bias term based on the predicted lemma to the logits of the factors. There is no additional transformer layer introduced with this option
+* `re-embedding`: After predicting a lemma, re-embed the lemma and add this new vector before predicting the factors
+* `lemma-dim-emb`: Controls the dimension of the re-embedded lemma when using the option `re-embedding`
+
+
+### Weight tying
+
+If using factors only on the source or target side but using a joint vocabulary, there are two different options to tie source and target embedding weights:
+
+1. Use combine option `concat` (If using factors only on the source side).
+2. Use combine option `sum`, and create "dummy" factors for the side initially without factors. This entails creating a factored vocabulary where the same number of factors are present as are on the side with meaningful factors. In the previous example, if we have the capitalization and subword factors on the source side, the target side would have five different dummy factors (they can all be in the same group). In the *lemma* section of the `.fsv` file we would just not put `_has_X` for any lemma.
```

Comment:
Would the term "the non-factorized side" be clearer?

Reply:
changed